plutus-ledger-api/testlib/PlutusLedgerApi/Test/V1/Value.hs

```diff
@@ -0,0 +1,95 @@
+-- example, if we only need to fill a single cell, we choose from 6 different candidates, and if we
+-- need to fill 5 cells, we choose from 11 candidates.
+--
+-- >>> map (\i -> (i, toCellCandidatesNumber i)) [1..13]
+-- [(1,6),(2,6),(3,6),(4,8),(5,11),(6,14),(7,18),(8,22),(9,27),(10,31),(11,36),(12,41),(13,46)]
+toCellCandidatesNumber :: Int -> Int
+toCellCandidatesNumber i = max 6 . floor @Double $ fromIntegral i ** 1.5
+
+-- | Generate a 'BuiltinByteString' by picking one of the predetermined ones, given a number of
+-- cells to fill (see 'toCellCandidatesNumber'). The idea is that we want to occasionally generate
+-- the same 'CurrencySymbol' or 'TokenName' for different 'Value's to have decent test coverage,
+-- hence to make name clashing more likely we pick from a predetermined set of
+-- 'BuiltinByteString's. Otherwise the chance of generating the same 'BuiltinByteString' for two
+-- different 'Value's would be virtually zero.
+genShortHex :: Int -> Gen BuiltinByteString
+genShortHex i =
+    toBuiltin . Base16.encode . BS8.pack . show <$> elements [0 .. toCellCandidatesNumber i]
+
+-- | Annotate each element of the give list with a @name@, given a function turning
+-- 'BuiltinByteString' into names.
+uniqueNames :: Eq name => (BuiltinByteString -> name) -> [b] -> Gen [(name, b)]
+uniqueNames wrap ys = do
+    let len = length ys
+    -- We always generate unique 'CurrencySymbol's within a single 'Value' and 'TokenName' within a
+    -- single 'CurrencySymbol', because functions over 'Value' don't handle duplicated names anyway.
+    -- Note that we can generate the same 'TokenName' within different 'CurrencySymbol's within the
+    -- same 'Value'.
+    xs <- uniqueVectorOf len $ wrap <$> genShortHex len
+    pure $ zip xs ys
+
+-- | The value of a 'TokenName' in a 'Value'.
+newtype FaceValue = FaceValue
+    { unFaceValue :: Integer
+    }
+
+instance Arbitrary FaceValue where
+    -- We want to generate zeroes often, because there's a lot of corner cases associated with them
+    -- and all non-zero numbers are handled pretty much the same anyway, so there isn't much point
+    -- in diversifying them as much as possible.
+    arbitrary = frequency
+        [ (2, pure $ FaceValue 0)
+        , (1, FaceValue . fromIntegral <$> arbitrary @Int)
+        ]
+
+-- | A wrapper for satisfying an @Arbitrary a@ constraint without implementing an 'Arbitrary'
+-- instance for @a@.
+newtype NoArbitrary a = NoArbitrary
+    { unNoArbitrary :: a
+    }
+
+-- | 'arbitrary' throws, 'shrink' neither throws nor shrinks.
+instance Arbitrary (NoArbitrary a) where
+    arbitrary = error "No such 'Arbitrary' instance"
+    shrink _ = []
+
+instance Arbitrary Value where
+    arbitrary = do
+        -- Generate values for all of the 'TokenName's in the final 'Value' and split them into a
+        -- list of lists.
+        faceValues <- fmap (map getNonEmpty) . multiSplit . map unFaceValue =<< arbitrary
```

Comment:
This is generating Values where no CurrencySymbol has an empty list of tokens, is that correct?  Do we actually want to exclude such things?  I can create them manually and they seem to behave as you'd expect (for example, applying `split` to such a thing gives me back two Values with a currency symbol with an empty list of tokens).  

I was wondering if this situation was prohibited by the ledger, but presumably it isn't since enforcing it would require quite a lot of checking.  Now I'm worrying about whether it effects equality.  Do we consider a value with a currency symbol which has no tokens to be equal to the same thing with that currency symbol missing?  The version of equality implemented in this PR doesn't consider them to be equal.  We should check whether equality of values is specified in the ledger specification or somewhere.

Reply:
> This is generating Values where no CurrencySymbol has an empty list of tokens, is that correct? 

Yes, and thank you for spotting that, it's really a generation bug. I've now fixed it, please review the [commit](https://github.com/input-output-hk/plutus/pull/5593/commits/4e89171e5176d5fcf49f80d8cd94992f5c5d58ee). I'll post some example distributions tomorrow or so.

> I originally thought that the version of == implemented [here](https://github.com/input-output-hk/plutus/blob/a0c01fbf08a441a7c16291c3ed69b550251eb40b/plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs#L449) didn't consider currency symbols with no tokens to be the same as absent currency symbols, but it does (and it says so here, which is a relief).

Yep, I didn't need to change anything in the implementation once I added generation of empty lists, tests kept being green (hope CI doesn't disprove that).

Reply:
Looks good.  Thanks!